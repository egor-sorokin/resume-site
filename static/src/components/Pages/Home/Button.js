import React from 'react';
import PropTypes from 'prop-types';


const Button = props => {
  const handleClick = () => {
    props.onClick('projects', true);
  };

  return (
    <div className="button button--white">
      <a
        href="#projects"
        className="button__link font-s-12"
        onClick={handleClick}
      >
        {props.children}
      </a>
    </div>
  );
};

Button.defaultProps = {
  children: [],
  onClick: () => {
  }
};

Button.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  onClick: PropTypes.func
};

export default Button;
