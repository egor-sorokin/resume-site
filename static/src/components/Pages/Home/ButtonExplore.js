import React from 'react';
import PropTypes from 'prop-types';


const ButtonExplore = ({ cssClasses, onClick, children }) => {
  const handleClick = () => {
    onClick('projects', true);
  };

  const classNames = `${cssClasses} button button--white`;

  return (
    <div className={classNames}>
      <a
        href="#projects"
        className="button__link font-s-12"
        onClick={handleClick}
      >
        {children}
      </a>
    </div>
  );
};

ButtonExplore.propTypes = {
  cssClasses: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  onClick: PropTypes.func,
};


ButtonExplore.defaultProps = {
  cssClasses: '',
  children: [],
  onClick: () => {
  },
};

export default ButtonExplore;
