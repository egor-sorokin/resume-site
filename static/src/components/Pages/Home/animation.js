import { Power2 } from 'gsap';


export const attachShowAnimation = (targetTween) => {
  targetTween
    .to('body', 0, { overflowY: 'hidden' })
    .fromTo(
      '.home', 0.8,
      { backgroundColor: 'rgba(231, 231, 231, 0.6)', ease: Power2.easeOut },
      { backgroundColor: 'rgba(34, 32, 29, 1)', ease: Power2.easeOut },
      0.35,
    )
    .fromTo(
      '.home .home__name', 0.5,
      { x: '50%', opacity: 0, ease: Power2.easeOut },
      { x: '0%', opacity: 1, ease: Power2.easeOut },
      1,
    )
    .fromTo(
      '.home .switcher .line--top', 0.5,
      { y: '-100%', ease: Power2.easeOut },
      { y: '0%', ease: Power2.easeOut },
      '-=0.3',
    )
    .fromTo(
      '.home .line--bottom', 0.5,
      { y: '100%', ease: Power2.easeOut },
      { y: '0%', ease: Power2.easeOut },
      '-=0.4',
    )
    .fromTo(
      '.home .switcher__link', 0.1,
      { x: '100%', opacity: 0, ease: Power2.easeOut },
      { x: '-50%', opacity: 1, ease: Power2.easeOut },
      '-=0.1',
    )
    .fromTo(
      '.home__button-explore', 0.5,
      { x: '0%', opacity: 0, ease: Power2.easeOut },
      { x: '-50%', opacity: 1, ease: Power2.easeOut },
    )
    .to('body', 0, { overflowY: 'initial' });
};
