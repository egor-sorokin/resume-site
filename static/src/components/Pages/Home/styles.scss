@import '../../../styles/base/_vars.scss';

.banner {
  position: relative;

  min-height: 100vh;

  background-color: $dune;
}

.home {
  &__name {
    display: flex;
    flex-direction: row;

    position: absolute;
    left: 10%;
    top: 50%;

    transform: translateY(-50%);
  }

  .line {
    position: absolute;
    right: 4%;

    height: 40vh;
    width: 1px;
    background-color: $mercuryLight;

    transform: translateX(-50%);

    &--bottom {
      bottom: 0;
    }
    &--top {
      top: 0;
    }
  }
}

.button {
  position: absolute;
  bottom: 40px;
  left: 50%;
  z-index: 2;

  cursor: pointer;

  transform: translateX(-50%);

  &:after {
    content: '';
    display: block;

    height: 1px;

    transform: translateX(0) translateZ(0) scaleX(1);
    transform-origin: left;

    transition: width $defaultTiming $defaultEasing,
    background-color $defaultTiming $defaultEasing,
    transform $defaultTiming $defaultEasing;
  }

  &__link {
    letter-spacing: 4px;

    transition: letter-spacing $defaultTiming $defaultEasing,
    color $defaultTiming $defaultEasing;
  }

  &--white {
    &:after {
      width: 55px;
      background-color: $mercuryLight;
    }
    .button__link {
      color: $mercuryLight;
    }
  }
  &--dark {
    &:after {
      width: 30px;
      background-color: $dune;
    }
    .button__link {
      color: $dune;
    }
  }
}

.button:hover {
  &:after {
    transform: translateX(6px) translateZ(0) scaleX(1.2);
  }

  .button__link {
    letter-spacing: 4.4px;
  }
}

.button--white:hover {
  &:after {
    background-color: $mercury;
  }
  .button__link {
    color: $mercury;
    letter-spacing: 4.4px;
  }
}

.button--dark:hover {
  &:after {
    background-color: $duneLight;
  }
  .button__link {
    color: $duneLight;
    letter-spacing: 5.4px;
  }
}