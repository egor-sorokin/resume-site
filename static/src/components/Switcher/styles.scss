@import '../../styles/base/_vars.scss';

.switcher {
  position: absolute;
  right: 0;
  top: 0;

  width: 4vw;
}

.switcher__line {
  height: 44vh;
  width: 1px;

  background-color: $dune;
  opacity: 0.3;
}

.switcher__link {
  display: flex;

  height: 12vh;

  align-items: center;
}

.switcher--white {
  .switcher__line {
    background-color: $mercury;
    opacity: 0.6;
  }
  .link-stretched {
    color: $mercuryLight;
  }
}
