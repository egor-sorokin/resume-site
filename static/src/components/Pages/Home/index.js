import React, {Component} from 'react';
import PropTypes from 'prop-types';
import scrollToY from 'scroll-to-y';
import ButtonExplore from './ButtonExplore';
import LinkStretched from '../../LinkStretched/index';
import withFetching from '../../../utils/api';
import {URL_PATH_PERSON_DATA, BUTTON_EXPLORE, BUTTON_ABOUT} from '../../../constants/index';
import './styles.scss';


class Home extends Component {
  scrollToProjects = (section, scrollTo) => {
    if (scrollTo) {
      scrollToY(
        document.getElementById(section).offsetTop,
        500,
        'easeInOutQuint'
      );
    }
  };


  clickAboutButton = () => {
    this.props.toggleAboutSection();
  };


  render() {
    const {data, isFetching, error} = this.props;
    const personData = data.personData || {};

    if (error) {
      return <div><p>{error}</p></div>;
    }

    if (isFetching) {
      return <div><p>Loading...</p></div>;
    }

    return (
      <div>
        <section id="home" className="home banner text-c-mercury-light">
          <h3 className="home__name font-s-36">
            {personData.first_name + ' ' + personData.last_name}
          </h3>

          <ButtonExplore
            onClick={this.scrollToProjects}
          >{BUTTON_EXPLORE.text}</ButtonExplore>

          <div className="switcher switcher--white">
            <div className="switcher__line line--top"/>
            <LinkStretched
              onClick={this.clickAboutButton}
              cssClasses="switcher__link font-s-12-secondary text-c-dune"
              text={BUTTON_ABOUT.text}
            ></LinkStretched>
            <div className="switcher__line line--bottom"/>
          </div>
        </section>
      </div>
    );
  }
}

Home.propTypes = {
  data: PropTypes.shape({}),
  isFetching: PropTypes.bool,
  error: PropTypes.string,
  toggleAboutSection: PropTypes.func
};

Home.defaultProps = {
  data: {},
  isFetching: false,
  error: '',
  toggleAboutSection: () => {
  }
};


export default withFetching(URL_PATH_PERSON_DATA, Home);
