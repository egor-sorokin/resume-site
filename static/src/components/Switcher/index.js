import React from 'react';
import PropTypes from 'prop-types';

import LinkStretched from '../LinkStretched/index';
import './styles.scss';


const Switcher = ({ url, switcherLink, onClick }) => {
  let link;
  const { type, text } = switcherLink;

  if (text && text.toLowerCase() === 'about') {
    link = (
      <LinkStretched
        cssClasses="switcher__link font-s-12-secondary text-c-mercury-light"
        text={text}
        onClick={onClick}
      />
    );
  } else if (url) {
    link = (
      <LinkStretched
        cssClasses="switcher__link font-s-12-secondary text-c-dune"
        text={text}
        onClick={onClick}
        url={url}
        target="_blank"
      />
    );
  }

  return (
    <div className={`switcher switcher--${type}`}>
      { link ? (
        <React.Fragment>
          <div className="switcher__line line line--top" />
          {link}
          <div className="switcher__line line line--bottom" />
        </React.Fragment>
      ) : (
        <div className="switcher__line line line--full" />
      )}
    </div>
  );
};

Switcher.propTypes = {
  onClick: PropTypes.func,
  url: PropTypes.string,
  switcherLink: PropTypes.shape({
    type: PropTypes.string,
    text: PropTypes.string,
  }),
};

Switcher.defaultProps = {
  onClick: () => {
  },
  switcherLink: {},
  url: '',
};

export default Switcher;
