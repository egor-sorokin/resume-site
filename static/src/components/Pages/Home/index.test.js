// eslint-disable-next-line
import {shallow, render, mount} from 'enzyme';
import React from 'react';
import Home from './index';
import ButtonExplore from './ButtonExplore';
import Switcher from '../../Switcher/index';


describe('<Home />', () => {
  let component;
  let spy;
  const isFetching = false;
  const person = {
    first_name: 'a',
    last_name: 'b',
  }
  const toggleAboutSection = jest.fn(() => {
  });
  const tweenline = {
    play: () => {
    }
  };


  describe('base tests', () => {
    beforeEach(() => {
      component = shallow(<Home person={person} isFetching={false}/>);
    });


    it('renders without crashing', () => {
      expect(component).toBeDefined();
    });


    it('matches snapshot', () => {
      expect(component).toMatchSnapshot();
    });


    it('renders 1 components', () => {
      console.log(component.debug());
      expect(component.children()).toHaveLength(3);
    });
  });


  describe('lifecycle tests', () => {
    beforeEach(() => {
      component = '';
    });


    afterEach(() => {
      spy.mockClear();
    });


    it('calls componentDidMount', () => {
      spy = jest.spyOn(Home.prototype, 'componentDidMount');
      // eslint-disable-next-line
      component = shallow(<Home person={person} isFetching={false}/>);
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });


  describe('props tests', () => {
    it('clicks on about button', () => {
      component = mount(<Home person={person} isFetching={false} toggleAboutSection={toggleAboutSection}/>);

      expect(component.find(<Switcher />)).toBeTruthy();
      expect(component.find(<ButtonExplore />)).toBeTruthy();
    });
  });


  describe('inner functions tests', () => {
    afterEach(() => {
      spy.mockClear();
    });


    it('calls play an animation', () => {
      spy = jest.spyOn(tweenline, 'play');
      // eslint-disable-next-line
      component = shallow(<Home person={person} isFetching={false}/>);
      component.setState({homeTween: tweenline});
      component.state().homeTween.play();
      expect(spy).toHaveBeenCalledTimes(1);

      spy.mockClear();
    });
  });
});


describe('<ButtonExplore />', () => {
  let component;
  const section = 'projects';
  const onClick = jest.fn(() => {
  });


  describe('base tests', () => {
    beforeEach(() => {
      component = shallow(<ButtonExplore />);
    });


    it('renders without crashing', () => {
      expect(component).toBeDefined();
    });


    it('matches snapshot', () => {
      expect(component).toMatchSnapshot();
    });


    it('renders 1 components', () => {
      expect(component.children()).toHaveLength(1);
    });
  });


  describe('inner functions tests', () => {
    it('clicks on explore button', () => {
      component = mount(<ButtonExplore onClick={onClick}/>);

      const exploreButton = component.find('a');
      expect(exploreButton).toBeTruthy();

      exploreButton.simulate('click', section);

      expect(onClick).toHaveBeenCalled();
    });
  });
});
