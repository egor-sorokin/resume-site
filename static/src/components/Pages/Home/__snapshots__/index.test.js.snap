// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonExplore /> base tests matches snapshot 1`] = `
<div
  className=" button button--white"
>
  <a
    className="button__link font-s-12"
    href="#projects"
    onClick={[Function]}
  />
</div>
`;

exports[`<Home /> base tests matches snapshot 1`] = `
<section
  className="banner home text-c-mercury-light"
  id="home"
>
  <h3
    className="home__name font-s-36"
  >
    a b
  </h3>
  <ButtonExplore
    cssClasses="home__button-explore"
    onClick={[Function]}
  >
    Explore
  </ButtonExplore>
  <Switcher
    onClick={[Function]}
    switcherLink={
      Object {
        "text": "About",
        "type": "white",
      }
    }
    url=""
  />
</section>
`;
